Update endpoints for blurhash implementation.

Frinksy opened the pull request:
Add blurhash to profile and avatar endpoints.

Resolves: https://github.com/ruma/ruma/issues/450

Frinksy committed:
Update endpoints for blurhash implementation.

Add blurhash to profile and avatar endpoints.

jplatte commented on crates/ruma-client-api/src/r0/profile/get_avatar_url.rs:
```diff
@@ -30,7 +30,16 @@ ruma_api! {
+        /// This uses the unstable prefix in
+        /// [MSC2448](https://github.com/matrix-org/matrix-doc/pull/2448).
+        #[serde(skip_serializing_if = "Option::is_none")]
+        #[cfg_attr(docsrs, doc(cfg(feature = "unstable-pre-spec")))]
```
The `doc(cfg)` attribute is there, but the regular `cfg` attribute (without `cfg_attr`) is missing, to actually feature-gate this field.

Frinksy replied:
I've added the regular `cfg` attributes to all the fields I've added.
They compile fine in CI and `cargo clippy --all-features`. However `cargo check --features full` (or `cargo check --features client-api`) fails. I'm not sure why that would be the case.

The error message is:
```
error[E0425]: cannot find value `blurhash` in this scope
  --> crates/ruma-client-api/src/r0/profile/set_avatar_url.rs:49:13
   |
49 |         pub blurhash: Option<&'a str>,
   |             ^^^^^^^^ not found in this scope

error[E0560]: struct `set_avatar_url::IncomingRequest` has no field named `blurhash`
  --> crates/ruma-client-api/src/r0/profile/set_avatar_url.rs:49:13
   |
49 |         pub blurhash: Option<&'a str>,
   |             ^^^^^^^^ `set_avatar_url::IncomingRequest` does not have this field
   |
   = note: available fields are: `user_id`, `avatar_url`
   ```

jplatte replied:
Oh, I need to look into this. Thanks for noticing.

DevinR528 replied:
Is this still a problem, it almost seems like it's a macro issue but that doesn't make any sense since it works for some features but not others.

jplatte replied:
Have not checked yet.

DevinR528 replied:
I figured it out #659

jplatte commented on crates/ruma-client-api/src/r0/profile/get_avatar_url.rs:
```diff
@@ -45,7 +54,7 @@ impl<'a> Request<'a> {
-    pub fn new(avatar_url: Option<MxcUri>) -> Self {
-        Self { avatar_url }
+    pub fn new(avatar_url: Option<MxcUri>, blurhash: Option<String>) -> Self {
+        Self { avatar_url, blurhash }
```
This constructor should not have a second argument, it should just set `blurhash` to `None` (under the pre-spec flag, otherwise it doesn't exist).

jplatte commented on crates/ruma-client-api/src/r0/profile/set_avatar_url.rs:
```diff
@@ -75,7 +88,7 @@ mod tests {
                     .body(&[] as &[u8]).unwrap(),
             ).unwrap(),
-            IncomingRequest { user_id, avatar_url: None } if user_id == "@foo:bar.org"
+            IncomingRequest { user_id, avatar_url: None, blurhash: None} if user_id == "@foo:bar.org"
```
You can use `, ..` instead of `, blurhash: None` here to make the test compile both with and without `unstable-pre-spec`.

jplatte reviewed (commented):
Comments on get_avatar_url apply to the other two modified modules as well.

jplatte commented:
Changed the base branch to main since this can (and should) be done in a backwards-compatible manner. You can `git rebase -i main` your branch and remove the unrelated commits as part of that to synchronize this PR to the `main` branch.

Frinksy committed:
Feature-gate blurhash and remove from constructor.

iinuwa reviewed (changes requested):
(Sorry for the late response.)

It looks like the changes to `GET /_matrix/federation/v1/query/profile` are missing. The previous implementation also only implemented the blurhash response field for `PUT /_matrix/media/r0/upload`, but not the `generate_blurhash` request query parameter.

The types are also missing for the `m.sticker` and `m.room.member` events.

Frinksy committed:
Add more blurhash fields.

Add `blurhash` fields to `GET /_matrix/federation/v1/query/profile`
and `m.room.member`.
Add `generate_blurhash` field to `PUT /_matrix/media/r0/upload`.

Frinksy commented:
> The types are also missing for the `m.sticker` and `m.room.member` events.

I've added the missing field for `m.room.member`, but I think `m.sticker` has already been dealt with. `m.sticker` uses `ruma_events::room::ImageInfo`, which already has a `blurhash` field.

jplatte commented on crates/ruma-client-api/src/r0/media/create_content.rs:
```diff
@@ -26,6 +26,16 @@ ruma_api! {
+        #[cfg(feature = "unstable-pre-spec")]
+        #[cfg_attr(docsrs, doc(cfg(feature = "unstable-pre-spec")))]
+        #[serde(rename = "xyz.amorgan.blurhash")]
+        pub generate_blurhash: &'a bool,
```
Trivially `Copy`-able types like bool or `js_int::UInt` shouldn't be borrowed.

jplatte commented on crates/ruma-client-api/src/r0/profile/set_avatar_url.rs:
```diff
@@ -75,7 +90,7 @@ mod tests {
                     .body(&[] as &[u8]).unwrap(),
             ).unwrap(),
-            IncomingRequest { user_id, avatar_url: None } if user_id == "@foo:bar.org"
+            IncomingRequest { user_id, avatar_url: None, ..} if user_id == "@foo:bar.org"
```
rustfmt doesn't know how to format all macro contents unfortunately. Can you add a space between `..` and `}` please?

jplatte commented on crates/ruma-client-api/src/r0/profile/set_avatar_url.rs:
```diff
@@ -87,7 +102,7 @@ mod tests {
                     .unwrap(),
             ).unwrap(),
-            IncomingRequest { user_id, avatar_url: None } if user_id == "@foo:bar.org"
+            IncomingRequest { user_id, avatar_url: None, ..} if user_id == "@foo:bar.org"
```
Same here.

Frinksy committed:
No longer borrow a bool and fix formatting.

DevinR528 reviewed (approved):
Everything looks good to me (as long as the other review is done, which it looks like you got it all)! Good catch with the gated fields that was a tricky one.

iinuwa reviewed (approved):
> I think m.sticker has already been dealt with. m.sticker uses ruma_events::room::ImageInfo, which already has a blurhash field.

Ah, yes. My bad.

This looks good to me! I haven't looked at the `#[cfg]` attribute issue in depth, but I assume we'll need to wait for #659 to be merged before merging this one.

jplatte commented:
#659 has been merged and CI adjusted to also check most things without unstable-pre-spec. This just needs a rebase now.
